### DISCUSSION TITLE
Control an external LCD during Circle boot.

### DISCUSSION THREAD

**Discussion by Newdrake** [rsta2/circle#522] [Development]
I would like to know if it is possible to modify the original circle boot code, that is the code that produces the rainbow color on the HDMI screen, to turn on an external LCD connected to the GPIO.
I ask this because in our application we have to wait about 10 seconds before seeing the LCD start writing something.
Is it possible to have this source code to insert the LCD activation?

Thanks a lot for your support!

**Comment by rsta2** [comment]
The rainbow screen is displayed by the firmware. It can be switched off, but not replaced. I think you can expect a GPIO-connected LCD display to display something after about two seconds, after Circle gets control from the firmware. The initialization of the member classes of `CKernel` must be reordered for this purpose. May I ask, what type of LCD display is it?

**Reply by Newdrake** [reply]
How to switch off the rainbow?

The LCD is a standard 20x4 chars, with 8bit parallel interface+W+E.
Is driven through a shift register, so I'm using the SPIO driver of circle to send commands to it.
After the boot I need to switch it on and visualize initialization string.
After the rainbow I have the sequence of bootstrap step like in the attached image, when I can start to control my LCD after the first row?
![image](https://github.com/user-attachments/assets/03e8a692-9a03-457e-b0c1-1b2df0466574)
In which source code I can insert my LCD initialization?

**Reply by Newdrake** [reply]
In my circle application I saw this timing, approximately:

0:00  Power on
5:46  Rainbow screen
8:96  Logger: Circle 44.2 ..
10:65  LCD on
Do you think this can be improved?

**Reply by rsta2** [reply]
The rainbow screen can be switched of with `disable_splash=1` in *config.txt*. Also `force_eeprom_read=0` and `disable_poe_fan=1` may help to reduce the boot time. See [this](https://www.raspberrypi.com/documentation/computers/config_txt.html#disable_splash) for more info.

Do you mean the `CSPIMaster` driver with `SPIO`? If so you should move the `CSPIMaster` to the first position in the list of member objects in `CKernel` and your LCD driver to the second. Then initialize the SPI master first and your LCD display second. You should be able to write to it then. In any case initialize the USBHCI member, the EMMC driver and the multi-core support after your display.

I did not expect these times. There is not much room for improvements between power on and 8:96 (see above). You also can use the cutdown firmware with `gpu_mem=16` in *config.txt* and compress your kernel image with `GZIP_KERNEL=1` in *Config.mk* (before build). But the firmware needs some time to tune the SDRAM timings. I think this takes most of the boot time.

**Reply by Newdrake** [reply]
Tanks a lot for this suggestion, I'll try all of these and let you know after how much time reduction I obtain,
Regards.

**Reply by Newdrake** [reply]
After many test and try to use the maximum number of boot time reduction parameter, I obtain a gain of about 2:20s, not bad.
This is the list of used switch:
initial_turbo=15
disable_splash=1
force_eeprom_read=0
disable_poe_fan=1
gpu_mem=16
ignore_lcd=1
disable_touchscreen=1
disable_fw_kms_setup=1
camera_auto_detect=0
display_auto_detect=0
enable_uart=0

So finally I obtain:
6:75s Logger: Circle 44.2 ..

Do you know a switch/variable to force the boot device or the image file? Without searching around devices like eporm, EMMC, ..

Thanks a lot again!!

**Reply by rsta2** [reply]
Not bad! The boot configuration on the CM4 can be changed [like that](https://www.raspberrypi.com/documentation/computers/compute-module.html#modify-the-bootloader-configuration). You are welcome.
